src/Features/Core/Portable/InheritanceMargin/Finders/BaseTypeSymbolsFinder.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.FindSymbols.FindReferences;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InheritanceMargin.Finders
+{
+    internal class BaseTypeSymbolsFinder : InheritanceSymbolsFinder
+    {
+        public static readonly BaseTypeSymbolsFinder Instance = new();
+
+        protected override Task<ImmutableArray<ISymbol>> GetAssociatedSymbolsAsync(
+            ISymbol symbol,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)symbol;
+            if (namedTypeSymbol.BaseType == null)
+            {
+                // AllInterfaces are topologically sorted by default.
+                return Task.FromResult(namedTypeSymbol.AllInterfaces.CastArray<ISymbol>());
+            }
+
+            // Calculate indegree for all the symbols
+            using var _ = PooledDictionary<ISymbol, HashSet<ISymbol>>.GetInstance(out var indegreeSymbolsMapBuilder);
```

Comment:
should we be using the MetadataUnifying comparer here?  I think we probably do want to do that right?

Reply:
The goal is to calculate like the 'symbol' before a symbol (the indegree map)
i.e.
```
interface IA { }
interface IB { }
class A : IA { }
class B : A, IB, IA { }
```
The graph looks like
![image](https://user-images.githubusercontent.com/24360909/161686321-c4a2c5d6-8dcb-4da5-88b5-ed67037d4ae5.png)
So that we have a dictionary like
{
    "IA" : ["A", "B"],
    "IB" : ["B"],
    "A" : ["B"],
    "B" : []
}

to do a topological sorting later.
I feel both should be fine MetadataUnifying comparer and the default dictionary should be fine here.
If it is going to create our own pool here then probably I could create a PooledDictionary with MetadataUnifying comparer and use it all the place.

Reply:
Fixed. Right now it is using a pooled object with MetadataUnifying comparer